### PR TITLE
Avoid arithmetic overflow in avg() function

### PR DESCRIPTION
**Pull request by jeffreylovitz**
This PR adds logic to the `avg()` function such that if its tracked total is going to overflow, it switches to an incremental algorithm that avoids this problem.

It is drawn from https://github.com/RedisTimeSeries/RedisTimeSeries/pull/918

### CODE REVIEW

File: tests/unit/test_aggregate_functions.cpp [hunk 1]
```diff
@@ -243,3 +243,40 @@ TEST_F(AggregateTest, StDevTest) {
...
+	SIValue expected_outcome = SI_DoubleVal((DBL_MAX / 2) + (1 / 2));
+	ASSERT_EQ(res.doubleval, expected_outcome.doubleval);
+
+	// third value is 2
+	avg->op.children[0] = two;
+	AR_EXP_Aggregate(avg, NULL);
+
+	// test average of DBL_MAX, 1, and 2
+	res = AR_EXP_Finalize(avg, NULL);
+	expected_outcome = SI_DoubleVal((DBL_MAX / 2) + (1 / 2) + (1 / 3));
```
Comment (swilly22) [inline tests/unit/test_aggregate_functions.cpp:274]: `expected_outcome = SI_DoubleVal(DBL_MAX/3 + 1/3 + 2/3);` ?

File: tests/unit/test_aggregate_functions.cpp [hunk 2]
```diff
@@ -243,3 +243,40 @@ TEST_F(AggregateTest, StDevTest) {
 	AR_EXP_Free(stdevp);
 }
 
+TEST_F(AggregateTest, AverageOverflowTest) {
```
[the remark this answers is not in the document]
Comment (swilly22) [reply tests/unit/test_aggregate_functions.cpp:246]: I think switching to a Python in this case will ease on writing multiple or event random tests

File: src/arithmetic/aggregate_funcs/agg_funcs.c [hunk 3]
```diff
@@ -73,32 +73,58 @@ AggregateResult AGG_SUM(SIValue *argv, int argc) {
...
+	// if we're uniquing inputs, return early if this value has already been seen
+	if(ctx->hashSet && Set_Add(ctx->hashSet, si_val) == false) return AGGREGATE_OK;
+
+	avg_ctx->count ++; // increment the count
+
+	// if we've already overflowed or adding the current value
+	// will cause us to overflow, use the incremental averaging algorithm
+	if(unlikely(avg_ctx->overflow || // already reached overflow
+				(signbit(avg_ctx->total) == signbit(v) && // each value has same sign
+				 (fabs(avg_ctx->total) > (DBL_MAX - fabs(v)))))) { // about to overflow
```
Comment (swilly22) [inline src/arithmetic/aggregate_funcs/agg_funcs.c:99]: we're getting quite a lot of `Codecov` complaints here, let make sure each line here is covered.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/unit/test_aggregate_functions.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply tests/unit/test_aggregate_functions.cpp:246]: The benefit to writing this as a unit test is that we can directly pass `DBL_MAX` - using Python would force us to switch to unwieldy large numbers (and `DBL_MAX` is technically implementation-defined, though in reality consistent).

File: src/arithmetic/aggregate_funcs/agg_funcs.c [hunk 4]
```diff
@@ -73,32 +73,58 @@ AggregateResult AGG_SUM(SIValue *argv, int argc) {
...
 
 	_agg_AvgCtx *avg_ctx = ctx->private_ctx;
 
-	// If we're uniquing inputs, return early if this value has already been seen.
-	if(ctx->hashSet && Set_Add(ctx->hashSet, v) == false) return AGGREGATE_OK;
-
-	avg_ctx->count ++;
-	avg_ctx->total += SI_GET_NUMERIC(v);
+	// if we're uniquing inputs, return early if this value has already been seen
+	if(ctx->hashSet && Set_Add(ctx->hashSet, si_val) == false) return AGGREGATE_OK;
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/arithmetic/aggregate_funcs/agg_funcs.c:91]: I'm willing to make that change, but personally think that the current solution is preferable. Branch prediction should stop this line from causing any performance penalty, and having 2 function versions would result in more lines of code and an uglier function rax. I'm open to talking it through, though!

File: src/arithmetic/aggregate_funcs/agg_funcs.c (hunk 3 above)
Comment (jeffreylovitz) [reply src/arithmetic/aggregate_funcs/agg_funcs.c:99]: The test was broken before, fixed now!

File: src/arithmetic/aggregate_funcs/agg_funcs.c [hunk 5]
```diff
@@ -73,32 +73,58 @@ AggregateResult AGG_SUM(SIValue *argv, int argc) {
...
+
+	avg_ctx->count ++; // increment the count
+
+	// if we've already overflowed or adding the current value
+	// will cause us to overflow, use the incremental averaging algorithm
+	if(unlikely(avg_ctx->overflow || // already reached overflow
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/arithmetic/aggregate_funcs/agg_funcs.c:97]: This was fun to test!

`unlikely` gave an average 0.18% improvement with the branch never taken, and gave a 0.006% detriment when the branch is taken.

I think this is narrow enough that any change is sampling error, and `unlikely` doesn't help.

File: tests/unit/test_aggregate_functions.cpp [hunk 6]
```diff
@@ -243,3 +243,29 @@ TEST_F(AggregateTest, StDevTest) {
...
+	// values to average
+	AR_ExpNode *max = AR_EXP_NewConstOperandNode(SI_DoubleVal(DBL_MAX));
```
Comment (swilly22) [inline tests/unit/test_aggregate_functions.cpp:248]: Please replicate the test for `SI_LongVal`

**Review by swilly22 — CHANGES_REQUESTED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**